ras_backstage/controllers/controller.py

```diff
@@ -24,23 +30,95 @@ def build_url(service_name, config, proxy_path):
+
+    return info
+
+
+@translate_exceptions
+def sign_in(config, username, password):
+    oauth_svc = config.dependency['oauth2-service']
+    client_id = oauth_svc['client_id']
+    client_secret = oauth_svc['client_secret']
+
+    oauth_payload = {
+        'grant_type': 'password',
+        'username': username,
+        'password': password
     }
-    info.update(current_app.config.get('METADATA', {}))
 
-    if current_app.config.feature.report_dependencies:
-        info["dependencies"] = [{'name': name} for name in current_app.config.dependency.keys()]
+    oauth_url = '{}://{}:{}/api/v1/tokens/'.format(oauth_svc['scheme'], oauth_svc['host'], oauth_svc['port'])
+    payload = '&'.join(["{}={}".format(k, v) for k, v in oauth_payload.items()])
+    response = requests.post(oauth_url,
+                             auth=(client_id, client_secret),
+                             data=payload,
+                             headers={'Content-Type': "application/x-www-form-urlencoded"})
+    response.raise_for_status()
+    response_data = response.json()
+
+    token_expiry = datetime.now() + timedelta(seconds=int(response_data['expires_in']))
+    response_data['expires_at'] = token_expiry.timestamp()
+    # FIXME: remove hard-coded values
+    response_data['party_id'] = 'BRES'
+    response_data['role'] = 'internal'
+
+    jwt_secret = config['SECRET_KEY']
+    token = jwt.encode(response_data, jwt_secret, algorithm=JWT_ALGORITHM)
+
+    return {'token': token}
+
+
+def validate_jwt(encoded_jwt_token):
+    log.debug("Attempting to validate JWT token.")
+    try:
+        jwt_secret = current_app.config['SECRET_KEY']
+        jwt_token = jwt.decode(encoded_jwt_token, jwt_secret, algorithms=JWT_ALGORITHM)
+    except JWTError:
+        raise RasError("Failed to decode JWT token.", status_code=401)
+
+    # Commented out because we don't have a requirement for JWT expiry yet...
+    # now = datetime.now()
+    # expires_at = datetime.fromtimestamp(jwt_token['expires_at'])
+    #
+    # if now > expires_at:
+    #     raise RasError("Token has expired.", status_code=401)
+
+    log.debug("JWT token validated successfully.")
+    return jwt_token
+
+
+def jwt_required(request):
+    def wrapper(f):
+        @wraps(f)
+        def decorator(*args, **kwargs):
+
+            if not current_app.config.feature['validate_jwt']:
+                return f(*args, **kwargs)
+
+            try:
+                encoded_jwt_token = request.headers['authorization']
+            except KeyError:
+                raise RasError("No JWT token provided", status_code=401)
 
-    return make_response(jsonify(info), 200)
+            jwt_token = validate_jwt(encoded_jwt_token)
+            return f(*args, **kwargs, token=jwt_token)
+        return decorator
+    return wrapper
 
 
 @translate_exceptions
-def proxy_request(request, service, url):
+@jwt_required(request)
+def proxy_request(config, request, service, url, token=None):
```

Comment:
The variable 'token' is never used? Is there a reason for it?

Reply:
Not really, removed.